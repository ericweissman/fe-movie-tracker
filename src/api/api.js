export const fetchData = async (url) => {
  const response = await fetch(url);
  if(response.ok) {
    return response.json()
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}

export const postData = async (urlSuffix, data) => {
  const url = 'http://localhost:3000/api/users' + urlSuffix;
  const response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(data),
    headers: {
      'Content-Type': 'application/json'
    }
  })
  if(response.ok) {
    let data = response.json();
    return data
  } else {
    throw Error(`Error fetching, ${response.status}`)
  }
}