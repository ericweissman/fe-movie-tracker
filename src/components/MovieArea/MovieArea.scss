.movie-area {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  align-items: center;
  justify-items: center;
  grid-gap: 25px;
}