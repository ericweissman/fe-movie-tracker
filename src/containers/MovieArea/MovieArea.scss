.movie-area {
  padding: 15px 0;
  display: flex;
  align-items: center;
  justify-items: center;
  flex-wrap: wrap;
  justify-content: center;
}


