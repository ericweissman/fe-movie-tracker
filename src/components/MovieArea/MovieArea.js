import React, { Component } from "react";
import { connect } from "react-redux";
import MovieCard from "../MovieCard/MovieCard";
import { populateFavorites } from "../../actions";
import { postData, deleteData } from "../../api/api";
import { getFavorites } from "../../utils/helper";

export class MovieArea extends Component {
  handleFavorite = async (movie, user) => {
    if (user.favorites && !user.favorites.includes(movie.movie_id)) {
      await postData("/favorites/new", { ...movie, user_id: user.id });
      let userFavorites = await getFavorites(user.id);
      this.props.populateFavorites(userFavorites);
    } else {
      await postData("/favorites/new", { ...movie, user_id: user.id });
      let userFavorites = await getFavorites(user.id);
      this.props.populateFavorites(userFavorites);
    }
  };

  handleDelete = async (movie, user) => {
    const suffix = `/${user.id}/favorites/${movie.movie_id}`;
    await deleteData(suffix, { user_id: user.id, movie_id: movie.movie_id });
    let userFavorites = await getFavorites(user.id);
    this.props.populateFavorites(userFavorites);
  };

  render() {
    if (this.props.location.pathname === "/favorites" && this.props.user.favorites) {
      const favorites = this.props.movies.filter(movie => {
        return this.props.user.favorites.includes(movie.movie_id);
      });
      return favorites.map(movie => {
        return (
          <MovieCard
            key={movie.movie_id}
            movie={movie}
            handleFavorite={this.handleFavorite}
            handleDelete={this.handleDelete}
          />
        );
      });
    } else if (
      this.props.location.pathname === "/favorites" &&
      !this.props.user.favorites
    ) {
      return <div>NO FAVORITES</div>;
    }
    return this.props.movies.map(movie => {
      return (
        <MovieCard
          key={movie.movie_id}
          movie={movie}
          handleFavorite={this.handleFavorite}
          user={this.props.user}
          handleDelete={this.handleDelete}
        />
      );
    });
  }
}

export const mapStateToProps = state => ({
  movies: state.movies,
  user: state.user
});

export const mapDispatchToProps = dispatch => ({
  populateFavorites: favorites => dispatch(populateFavorites(favorites))
});

export default connect(mapStateToProps, mapDispatchToProps)(MovieArea);
