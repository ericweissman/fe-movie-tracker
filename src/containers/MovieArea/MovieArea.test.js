import React from "react"
import { shallow } from "enzyme"
import { mapStateToProps, MovieArea } from "./MovieArea"
import { handleFavorite } from "../MovieArea/MovieArea"
// import { getFavorites } from "../../utils/helper";
// import { populateFavorites } from "../../actions"
import * as api from "../../api/api"
jest.mock("../../api/api")

describe("MovieArea", () => {
  let wrapper
  let moviesMock = [
    {
      vote_count: 3078,
      movie_id: 1,
      video: false,
      vote_average: 6.9,
      title: "Aquaman",
      popularity: 805.109,
      poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
      original_language: "en",
      original_title: "Aquaman"
    },
    {
      vote_count: 3078,
      movie_id: 2,
      video: false,
      vote_average: 6.9,
      title: "Aquaman",
      popularity: 805.109,
      poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
      original_language: "en",
      original_title: "Aquaman"
    }
  ]
  let newFavMovie = {
    vote_count: 3078,
    movie_id: 3,
    video: false,
    vote_average: 6.9,
    title: "Aquaman",
    popularity: 805.109,
    poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
    original_language: "en",
    original_title: "Aquaman"
  }
  let userMock = { name: "Hill", id: 265, favorites: [1, 2] }
  let mockLocation = { pathname: "/favorites" }
  let handleFavorite = jest.fn()
  let mockData = {
    status: "success",
    message: "Movie was added to favorites",
    id: 3
  }
  let mockFavoriteIDs = [2, 3]
  let getFavorites = jest.fn(() => mockFavoriteIDs)
  let populateFavoritesMock = jest
    .fn()
    .mockImplementation(() => Promise.all(mockFavoriteIDs))

  it("should match the correct snapshot", () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe("handleFavorite", () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(
        <MovieArea
          location={mockLocation}
          user={userMock}
          movies={moviesMock}
          populateFavorites={populateFavoritesMock}
        />
      )
    })
    it.skip("should call postData, getFavorites, populateFavorites if the movie is not in the user favorites", async () => {
      handleFavorite(newFavMovie, userMock)
      api.postData = await jest.fn(() => mockData)
      let favs = await getFavorites(userMock.id)
      expect(wrapper.instance().props.populateFavorites).toHaveBeenCalledWith(
        ...favs
      )
    })
  })

  describe("mapStateToProps", () => {
    it("should return an object with the movies in an array", () => {
      const movies = [
        {
          vote_count: 3078,
          id: 297802,
          video: false,
          vote_average: 6.9,
          title: "Aquaman",
          popularity: 805.109,
          poster_path: "/5Kg76ldv7VxeX9YlcQXiowHgdX6.jpg",
          original_language: "en",
          original_title: "Aquaman"
        }
      ]
      const mockState = {
        movies,
        user: {}
      }
      const expected = { movies, user: {} }
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps).toEqual(expected)
    })
  })
})
