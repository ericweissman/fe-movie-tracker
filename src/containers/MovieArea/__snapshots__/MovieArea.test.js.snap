// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieArea should match the correct snapshot 1`] = `undefined`;
